### PR TITLE
Draw bubble backgrounds manually

### PR DESCRIPTION
**Pull request by JozefCeluch**
#### Problem

Drawable XMLs that we use for backgrounds are not drawn correctly on SDK version smaller than 23. 

The weight/height attributes of items in the layer-list are ignored and the items are stretched to match the parent size. This made it impossible to create small circles in the message layout and the welcome/login screen layouts.

#### Solution

Remove the XML drawables and draw the backgrounds directly in the onDraw method of the custom views.

#### Screenshots 

| Before | After |
| --- | --- |
| login_before | login_after |
| message_before | message_after |

### CODE REVIEW

File: Firebase/android/app/src/main/res/layout/merge_login_view.xml [hunk 1]
```diff
@@ -8,18 +8,19 @@
     android:layout_height="match_parent"
     android:orientation="vertical">
 
-    <FrameLayout
+    <com.novoda.bonfire.view.BubblyLinearLayout
```
Comment (rock3r) [inline Firebase/android/app/src/main/res/layout/merge_login_view.xml]: Why the switch from `FrameLayout` to `LinearLayout` with no changes to the contents? ⚠️

File: Firebase/android/app/src/main/java/com/novoda/bonfire/view/BubblyLinearLayout.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import android.content.res.TypedArray;
+import android.graphics.Canvas;
+import android.graphics.Paint;
+import android.graphics.RectF;
+import android.util.AttributeSet;
+import android.widget.LinearLayout;
+
+import com.novoda.bonfire.R;
+
+public class BubblyLinearLayout extends LinearLayout {
```
Comment (rock3r) [inline Firebase/android/app/src/main/java/com/novoda/bonfire/view/BubblyLinearLayout.java]: Shouldn't be a `LinearLayout` but rather a `FrameLayout` -- see https://github.com/novoda/spikes/pull/107/files#r67866404

Comment (rock3r) [inline Firebase/android/app/src/main/java/com/novoda/bonfire/view/BubblyLinearLayout.java]: Ok there's cases where this needs to be a `LinearLayout`, nvm

**Comment by rock3r** [comment]
I think it'd be way easier to create a `BubbleDrawable` instead, and use that for BGs...

File: Firebase/android/app/src/main/res/layout/merge_login_view.xml (hunk 1 above)
Comment (JozefCeluch) [inline Firebase/android/app/src/main/res/layout/merge_login_view.xml]: The other place where it is used has multiple views inside the layout so extended the layout that would work in both places

**Comment by JozefCeluch** [comment]
@rock3r It would be easier, and we had it like that but it didn't work on APIs smaller than 23 because the width and height of items in layer-list are ignored. The easiest solution for message bubble would be a 9 patch but that wouldn't work for the bubble in the login view anyway.

**Comment by rock3r** [comment]
Maybe I wasn't explaining as best as I could:

```java
public class BubblyDrawable extends Drawable {

  // ...

  @Override
  public void draw() {
    // Your drawing logic here
  }
}
```

Then you just `setBackground(new BubblyDrawable())` :)

**Comment by JozefCeluch** [comment]
Oh yeah, good point. But I would have to set it in the code instead of having everything in the xml, right?

**Comment by rock3r** [comment]
Yup, but it's EZ.

File: Firebase/android/app/src/main/java/com/novoda/bonfire/login/view/LoginView.java [hunk 3]
```diff
@@ -24,6 +25,7 @@ protected void onFinishInflate() {
         super.onFinishInflate();
         View.inflate(getContext(), R.layout.merge_login_view, this);
         loginButton = Views.findById(this, R.id.sign_in_button);
+        Views.findById(this, R.id.login_icon_frame_layout).setBackground(new BubblyDrawable(getContext()));
```
Comment (rock3r) [inline Firebase/android/app/src/main/java/com/novoda/bonfire/login/view/LoginView.java]: What about doing one thing at a time?

```java
View frame = Views.findById(this, R.id.login_icon_frame_layout);
frame.setBackground(new BubblyDrawable(getContext()));
```

File: Firebase/android/app/src/main/java/com/novoda/bonfire/view/BubblyDrawable.java [hunk 4]
```diff
@@ -75,4 +58,18 @@ private void drawBubble(Canvas canvas, int xPosition, int yPosition, int diamete
...
+    }
+
+    @Override
+    public void setColorFilter(ColorFilter colorFilter) {
+        paint.setColorFilter(colorFilter);
+    }
+
+    @Override
+    public int getOpacity() {
+        return PixelFormat.OPAQUE;
```
Comment (rock3r) [inline Firebase/android/app/src/main/java/com/novoda/bonfire/view/BubblyDrawable.java]: Should be `TRANSLUCENT` unless _every_ pixel within the drawable's bounds is 100% opaque! ⚠️

File: Firebase/android/app/src/main/java/com/novoda/bonfire/welcome/view/WelcomeView.java [hunk 5]
```diff
@@ -34,6 +35,7 @@ protected void onFinishInflate() {
         welcomeMessage = Views.findById(this, R.id.welcome_message_view);
         userAvatar = Views.findById(this, R.id.user_avatar);
         proceedButton = Views.findById(this, R.id.proceed_button);
+        Views.findById(this, R.id.welcome_sender_layout).setBackground(new BubblyDrawable(getContext()));
```
Comment (rock3r) [inline Firebase/android/app/src/main/java/com/novoda/bonfire/welcome/view/WelcomeView.java]: Same as before

**Comment by rock3r** [comment]
Couple of comments before LGTM

**Comment by rock3r** [comment]
LGTM now 👍

**Comment by Dorvaryn** [comment]
Should the TextView still be a custom view ? Seems like we could use a custom drawable as well.

**Comment by rock3r** [comment]
That makes sense too @Dorvaryn -- my bad for missing that out. Assumed it'd be in a separate PR, not sure why. Was tired I suppose :D

File: Firebase/android/app/src/main/java/com/novoda/bonfire/chat/view/ChatAdapter.java [hunk 6]
```diff
@@ -30,11 +33,17 @@ public void update(Chat chat, User user) {
 
     @Override
     public MessageViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+        MessageBubbleDrawable bubbleDrawable;
+        MessageView messageView;
         if (viewType == 0) {
```
Comment (rock3r) [inline Firebase/android/app/src/main/java/com/novoda/bonfire/chat/view/ChatAdapter.java]: Could I recommend extracting a `VIEW_TYPE_MESSAGE_OTHER_USERS` maybe? :) Also a `VIEW_TYPE_MESSAGE_THIS_USER` and use it here (`else if` it), so you have an `else` that can throw

File: Firebase/android/app/src/main/java/com/novoda/bonfire/view/MessageBubbleDrawable.java [hunk 7]
```diff
@@ -26,32 +27,15 @@
...
+    public MessageBubbleDrawable(Context context, @ColorRes int color, Gravity gravity) {
+        this.gravity = gravity;
         rect = new RectF();
         paint = new Paint();
         paint.setStyle(Paint.Style.FILL);
         paint.setAntiAlias(true);
-        resolveAttributes(context, attrs);
-        resolveDimensions(getResources());
-    }
+        paint.setColor(context.getResources().getColor(color));
```
Comment (rock3r) [inline Firebase/android/app/src/main/java/com/novoda/bonfire/view/MessageBubbleDrawable.java]: This, or use `setColorFilter()` to do the tinting:

```java
ColorFilter colorFilter = new PorterDuffColorFilter(myColor, PorterDuff.Mode.SRC_IN);
setColorFilter(colorFilter);
```

**Comment by tasomaniac** [comment]
Looks good but just 1 comment. Wouldn't this be very easy with 9-patch as well?

File: Firebase/android/app/src/main/java/com/novoda/bonfire/view/MessageBubbleDrawable.java (hunk 7 above)
Comment (JozefCeluch) [inline Firebase/android/app/src/main/java/com/novoda/bonfire/view/MessageBubbleDrawable.java]: Does setting a color have any disadvantages compared to setting a color filter?

**Comment by JozefCeluch** [comment]
@tasomaniac Yes, for the messages 9-patch would work

**Comment by rock3r** [comment]
@tasomaniac https://github.com/novoda/spikes/pull/107#issuecomment-227441858

File: Firebase/android/app/src/main/java/com/novoda/bonfire/view/MessageBubbleDrawable.java (hunk 7 above)
Comment (rock3r) [inline Firebase/android/app/src/main/java/com/novoda/bonfire/view/MessageBubbleDrawable.java]: Nah, it's just an FYI, I don't mind either way :)

**Comment by tasomaniac** [comment]
Oh, I read the comments but totally missed that part of @JozefCeluch 's comment. 👍

**Pull request merged**